### PR TITLE
Added a reporter

### PR DESCRIPTION
**Pull request by yannickl88**

### CODE REVIEW

File: src/Bundler/Pipeline/ContentPipeline.php [hunk 1]
```diff
@@ -137,9 +135,11 @@ public function push(array $dependencies, ReaderInterface $file_reader, File $ta
                     );
                 }
 
-                $this->logger->info('   - Emiting compile file for {name}', ['name' => $item->file->path]);
+                $reporter->reportFileState($item->file, ReporterInterface::STATE_BUILD);
```
Comment (iltar) [inline src/Bundler/Pipeline/ContentPipeline.php]: `STATE_BUILT`

File: src/Config/SimpleConfig.php [hunk 2]
```diff
@@ -198,4 +203,20 @@ public function getRunner(): RunnerInterface
 
         return $this->runner;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setReporter(ReporterInterface $reporter): void
```
Comment (iltar) [inline src/Config/SimpleConfig.php]: `replaceReporter`? `changeReporter`?

File: src/Report/ConsoleLoggingReporter.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+use Hostnet\Component\Resolver\File;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Reporter which outputs directly to the OutputInterface when a new output
+ * file has been reported.
+ *
+ * This will mimic a logger and can be used for deployment output.
+ */
+class ConsoleLoggingReporter implements ReporterInterface
```
Comment (iltar) [inline src/Report/ConsoleLoggingReporter.php]: final

File: src/Report/ConsoleReporter.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+namespace Hostnet\Component\Resolver\Report;
+
+use Hostnet\Component\Resolver\Config\ConfigInterface;
+use Hostnet\Component\Resolver\File;
+use Hostnet\Component\Resolver\Import\DependencyNodeInterface;
+use Symfony\Component\Console\Helper\Table;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ConsoleReporter implements ReporterInterface
```
Comment (iltar) [inline src/Report/ConsoleReporter.php]: final

File: src/Report/ConsoleReporter.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            return [];
+        }
+
+        $files = array_merge(...array_values($this->dependencies));
+        $names = array_values(array_unique(array_map(function (DependencyNodeInterface $dep) {
+            return $dep->getFile()->path;
+        }, $files)));
+        $reasons = array_combine($names, array_fill(0, count($names), []));
+
+        foreach ($this->dependencies as $input_file => $dependencies) {
```
Comment (iltar) [inline src/Report/ConsoleReporter.php]: $input_file can be removed

File: src/Report/ConsoleReporter.php [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            foreach ($dependencies as $dependency) {
+                foreach ($dependency->getChildren() as $child) {
+                    $reasons[$child->getFile()->path][] = sprintf(
+                        'Included by [%d] <fg=cyan>%s</>',
+                        array_search($dependency->getFile()->path, $names, true),
+                        $dependency->getFile()->path
```
Comment (iltar) [inline src/Report/ConsoleReporter.php]: `$dependency->getFile()` can be moved to a local var

File: test/Report/report.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+         Asset     Size     Status       
+ fixtures/a.js 1 kb -> 3 kb [build]      
+ fixtures/b.js     0 b      [up-to-date] 
+
+ [0]  fixtures/a.js 3 kb [build]
```
Comment (iltar) [inline test/Report/report.txt]: built

File: src/Report/NullReporter.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace Hostnet\Component\Resolver\Report;
+
+use Hostnet\Component\Resolver\File;
+
+/**
+ * Reporter does nothing.
+ */
+class NullReporter implements ReporterInterface
```
Comment (iltar) [inline src/Report/NullReporter.php]: final

File: src/Report/ReporterInterface.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+declare(strict_types=1);
+namespace Hostnet\Component\Resolver\Report;
+
+use Hostnet\Component\Resolver\File;
+
+interface ReporterInterface
+{
+    public const STATE_UP_TO_DATE = 'up-to-date';
+    public const STATE_FROM_CACHE = 'from-cache';
+    public const STATE_BUILD      = 'build';
```
Comment (iltar) [inline src/Report/ReporterInterface.php]: built

File: src/Bundler/PipelineBundler.php [hunk 10]
```diff
@@ -60,16 +60,24 @@ public function execute(ReaderInterface $reader, WriterInterface $writer): void
...
                     $require_file,
                     $output_require_file->getName(),
                     new StringReader($reader->read($require_file))
                 );
 
-                $writer->write($output_require_file, $this->runner->execute(RunnerType::UGLIFY, $item));
+                $content = $this->runner->execute(RunnerType::UGLIFY, $item);
+
+                $reporter->reportFileState($output_require_file, ReporterInterface::STATE_BUILT);
+                $reporter->reportFileSize($output_require_file, \strlen($content));
```
Comment (nicoschoenmaker) [inline src/Bundler/PipelineBundler.php]: Shouldn't we do a `reportFileContent` instead to avoid the `strlen` operation for a `NullReporter`?

**Pull request merged**